sycl/tools/sycl-trace/collector.cpp

```diff
@@ -8,21 +8,146 @@
 
 #include "xpti/xpti_trace_framework.h"
 
+#include <dlfcn.h>
+#include <iostream>
+#include <mutex>
+#include <string>
 #include <sycl/detail/spinlock.hpp>
 
 sycl::detail::SpinLock GlobalLock;
 
 bool HasZEPrinter = false;
-bool HasCUPrinter = false;
-bool HasPIPrinter = false;
-bool HasSYCLPrinter = false;
 
-void zePrintersInit();
-void zePrintersFinish();
+std::string getCurrentDSODir() {
+  auto CurrentFunc = reinterpret_cast<const void *>(&getCurrentDSODir);
+  Dl_info Info;
+  int RetCode = dladdr(CurrentFunc, &Info);
+  if (0 == RetCode) {
+    // This actually indicates an error
+    return "";
+  }
+
+  auto Path = std::string(Info.dli_fname);
+  auto LastSlashPos = Path.find_last_of('/');
+
+  return Path.substr(0, LastSlashPos);
+}
+
+class CollectorLibraryWrapper {
+  typedef void (*InitFuncType)();
+  typedef void (*FinishFuncType)();
+  typedef void (*CallbackFuncType)(uint16_t, xpti::trace_event_data_t *,
+                                   xpti::trace_event_data_t *, uint64_t,
+                                   const void *);
+  typedef void (*SetIndentLvlFuncType)(int);
+
+public:
+  CollectorLibraryWrapper(const std::string &LibraryName)
+      : MLibraryName(LibraryName){};
+  ~CollectorLibraryWrapper() { clear(); };
+
+  const std::string InitFuncName = "init";
+  const std::string FinishFuncName = "finish";
+  const std::string CallbackFuncName = "callback";
+  const std::string IndentFuncName = "setIndentationLevel";
+
+  bool initPrinters() {
+    const std::string CollectorDir = getCurrentDSODir() + "/" + MLibraryName;
```

Comment:
Based on the above, `getCurrentDSODir()` can return the empty string in certain error cases. If this happens, this directory path would be `/some_library_name` which could be a problem. Instead, the value returned by `getCurrentDSODir()` should be checked first.

Reply:
done, good catch, thanks